File: src/main/java/audioWfc/wfc/constraints/AscendingMelodySoftConstraint.java
package audioWfc.wfc.constraints;

import audioWfc.wfc.HigherValues;
import audioWfc.musicTheory.OctavedNote;
import audioWfc.wfc.Tile;
import audioWfc.wfc.constraints.concepts.SoftConstraint;

import static audioWfc.wfc.constraints.ConstraintUtils.ASCENDING_MELODY;

public class AscendingMelodySoftConstraint extends SoftConstraint<OctavedNote> {
    public AscendingMelodySoftConstraint(double factor){
        super(factor);
    }

    @Override
    public double weight(Tile<OctavedNote> item, HigherValues higherValues) {
        OctavedNote note = item.getValue();

        Tile<OctavedNote> prev = item.getPrev();
        Tile<OctavedNote> next = item.getNext();

        double out = 0d;

        if(prev.isCollapsed() && checkAscending(prev.getValue(), note)){
            out += factor;
        }

        if(next.isCollapsed() && checkAscending(note, next.getValue())){
            out += factor;
        }

        return out;
    }

    private boolean checkAscending(OctavedNote first, OctavedNote second){
        return second.MIDIValue() > first.MIDIValue();
    }


    @Override
    public String name() {
        return ASCENDING_MELODY;
    }

    @Override
    public String configText() {
        return "Bonus: " + factor;
    }
}

File: vite-project/src/wfc/constraints/AscendingMelodySoftConstraint.ts
import { OctavedNote } from "../../music_theory/Note"
import { Tile } from "../Tile"
import { SoftConstraint } from "./concepts/Constraint"
import { noteConstraintTypeToName } from "./constraintUtils"

export class AscendingMelodySoftConstraint extends SoftConstraint<OctavedNote>{
	name = noteConstraintTypeToName.get("AscendingMelodySoftConstraint") as string
	configText = () => `Bonus: ${this.bonus}`

	constructor(bonus: number) {
		super(bonus)
	}

	weight(tile: Tile<OctavedNote>): number {
		const note = tile.getValue()
		const prev = tile.getPrev()
		const next = tile.getNext()

		let out = 0

		if(prev.isCollapsed() && this.isAscending(prev.getValue(), note)) out += this.bonus
		if(next.isCollapsed() && this.isAscending(note, next.getValue())) out += this.bonus

		return out
	}

	private isAscending(first: OctavedNote, second: OctavedNote): boolean {
		return first.toMIDIValue() < second.toMIDIValue()
	}

}